tools/lsp/preview/ui/palette.rs

```diff
@@ -0,0 +1,351 @@
+        let (dc, url) = compile(
+            r#"
+global Other {
+    out property <color> color1: #1ff;
+    out property <color> color2: #2ff;
+    out property <color> color3: #3ff;
+}
+
+global Test {
+    in property <int> index;
+    out property <color> color1: #f0f;
+    out property <color> color2: Other.color2;
+    out property <color> color3 <=> Other.color3;
+    in property <color> color4;
+    out property <color> color5: index == 0 ? Other.color1 : Other.color2;
+}
+
+export component Main { }
+            "#,
+        );
+        let result = collect_palette_from_globals(&dc, &url, Vec::new());
+        assert_eq!(result.len(), 7);
+
+        compare(&result[0], "Other.color1", 0x11, 0xff, 0xff);
+        compare(&result[1], "Other.color2", 0x22, 0xff, 0xff);
+        compare(&result[2], "Other.color3", 0x33, 0xff, 0xff);
+
+        compare(&result[3], "Test.color1", 0xff, 0x00, 0xff);
+        compare(&result[4], "Test.color2", 0x22, 0xff, 0xff);
+        compare(&result[5], "Test.color3", 0x33, 0xff, 0xff);
+        compare(&result[6], "Test.color5", 0x11, 0xff, 0xff);
+    }
+
+    #[test]
+    fn test_globals_palettes_with_struct() {
+        let (dc, url) = compile(
+            r#"
+struct Colors {
+    color1: color,
+    color2: color,
+    color3: color,
+}
+
+global Test {
+    in property <int> index;
+
+    out property <Colors> palette: {
+        color1: #1ff,
+        color2: #2ff,
+        color3: #3ff,
+    };
+}
+
+export component Main { }
+            "#,
+        );
+        let result = collect_palette_from_globals(&dc, &url, Vec::new());
+        assert_eq!(result.len(), 3);
+
+        compare(&result[0], "Test.palette.color1", 0x11, 0xff, 0xff);
+        compare(&result[1], "Test.palette.color2", 0x22, 0xff, 0xff);
+        compare(&result[2], "Test.palette.color3", 0x33, 0xff, 0xff);
+    }
+
+    #[test]
+    fn test_globals_palettes_with_struct_conditional() {
+        let (dc, url) = compile(
+            r#"
+struct Colors {
+    color1: color,
+    color2: color,
+    color3: color,
+}
+
+global Test {
+    in property <int> index;
+
+    out property <Colors> _0: {
+        color1: #1ff,
+        color2: #2ff,
+        color3: #3ff,
+    };
+
+    out property <Colors> _1: {
+        color1: #111,
+        color2: #222,
+        color3: #333,
+    };
+
+    out property <Colors> palette: root.index == 0 ? root._0 : root._1;
+}
+
+export component Main { }
+            "#,
+        );
+        let result = collect_palette_from_globals(&dc, &url, Vec::new());
+        assert_eq!(result.len(), 9);
+
+        compare(&result[0], "Test.-0.color1", 0x11, 0xff, 0xff);
+        compare(&result[1], "Test.-0.color2", 0x22, 0xff, 0xff);
+        compare(&result[2], "Test.-0.color3", 0x33, 0xff, 0xff);
+
+        compare(&result[3], "Test.-1.color1", 0x11, 0x11, 0x11);
+        compare(&result[4], "Test.-1.color2", 0x22, 0x22, 0x22);
+        compare(&result[5], "Test.-1.color3", 0x33, 0x33, 0x33);
+
```

Comment:
This needs to be `_0` and `_1`, as the leading `-` makes the code invalid.

I am trying to understand why the normalization function does not work here...

Reply:
PR #8490 fixed this!